### PR TITLE
 [SPARK-34130][SQL] Impove preformace for char varchar padding and length check with StaticInvoke

### PR DESCRIPTION
**Pull request by yaooqinn**
### What changes were proposed in this pull request?

This could reduce the `generate.java` size to prevent codegen fallback which causes performance regression.

here is a case from tpcds that could be fixed by this improvement
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133964/testReport/org.apache.spark.sql.execution/LogicalPlanTagInSparkPlanSuite/q41/

The original case generate 20K bytes, we are trying to reduce it to less than 8k
### Why are the changes needed?

performance improvement as in the PR benchmark test, the performance  w/ codegen is 2~3x better than w/o codegen.

### Does this PR introduce _any_ user-facing change?

no

### How was this patch tested?

yes, it's a code reflect so the existing ut should be enough

cross-check with https://github.com/apache/spark/pull/31012 where the tpcds shall all pass

benchmark compared with master

```logtalk
================================================================================================
Char Varchar Read Side Perf
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Read with length 20, hasSpaces: false:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read string with length 20                         1571           1667          83         63.6          15.7       1.0X
read char with length 20                           1710           1764          58         58.5          17.1       0.9X
read varchar with length 20                        1774           1792          16         56.4          17.7       0.9X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Read with length 40, hasSpaces: false:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read string with length 40                         1824           1927          91         54.8          18.2       1.0X
read char with length 40                           1788           1928         137         55.9          17.9       1.0X
read varchar with length 40                        1676           1700          40         59.7          16.8       1.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Read with length 60, hasSpaces: false:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read string with length 60                         1727           1762          30         57.9          17.3       1.0X
read char with length 60                           1628           1674          43         61.4          16.3       1.1X
read varchar with length 60                        1651           1665          13         60.6          16.5       1.0X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Read with length 80, hasSpaces: true:     Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read string with length 80                         1748           1778          28         57.2          17.5       1.0X
read char with length 80                           1673           1678           9         59.8          16.7       1.0X
read varchar with length 80                        1667           1684          27         60.0          16.7       1.0X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Read with length 100, hasSpaces: true:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read string with length 100                        1709           1743          48         58.5          17.1       1.0X
read char with length 100                          1610           1664          67         62.1          16.1       1.1X
read varchar with length 100                       1614           1673          53         61.9          16.1       1.1X


================================================================================================
Char Varchar Write Side Perf
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Write with length 20, hasSpaces: false:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write string with length 20                        2277           2327          67          4.4         227.7       1.0X
write char with length 20                          2421           2443          19          4.1         242.1       0.9X
write varchar with length 20                       2393           2419          27          4.2         239.3       1.0X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Write with length 40, hasSpaces: false:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write string with length 40                        2249           2290          38          4.4         224.9       1.0X
write char with length 40                          2386           2444          57          4.2         238.6       0.9X
write varchar with length 40                       2397           2405          12          4.2         239.7       0.9X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Write with length 60, hasSpaces: false:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write string with length 60                        2326           2367          41          4.3         232.6       1.0X
write char with length 60                          2478           2501          37          4.0         247.8       0.9X
write varchar with length 60                       2475           2503          24          4.0         247.5       0.9X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Write with length 80, hasSpaces: true:    Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write string with length 80                        9367           9773         354          1.1         936.7       1.0X
write char with length 80                         10454          10621         238          1.0        1045.4       0.9X
write varchar with length 80                      18943          19503         571          0.5        1894.3       0.5X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
Write with length 100, hasSpaces: true:   Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write string with length 100                      11055          11104          59          0.9        1105.5       1.0X
write char with length 100                        12204          12275          63          0.8        1220.4       0.9X
write varchar with length 100                     21737          22275         574          0.5        2173.7       0.5X

```

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Since this is a performance PR, could you provide a benchmark result, @yaooqinn ?

Or, do we need to have one benchmark suite on CHAR/VARCHAR, @cloud-fan ?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+package org.apache.spark.sql.catalyst.util;
+
+import org.apache.spark.unsafe.types.UTF8String;
+
+public class CharVarcharCodegenUtils {
+  private static final UTF8String SPACE = UTF8String.fromString(" ");
+
+  public static UTF8String lengthCheck(UTF8String inputStr, int limit) {
+    if (inputStr == null) return null;
+    if (inputStr.numChars() > limit) {
```
Comment (srowen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:27]: Not sure if it helps code size or speed, but you can skip the if/return above and add "inputStr != null &&" here for the same result.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 2]
```diff
@@ -200,69 +212,60 @@ object CharVarcharUtils extends Logging {
...
-            LessThanOrEqual(Length(expr), Literal(length)),
-            expr,
-            If(
-              GreaterThan(Length(trimmed), Literal(length)),
-              raiseError("varchar", length),
-              StringRPad(trimmed, Literal(length))))
-        } else {
-          If(GreaterThan(Length(expr), Literal(length)), raiseError("varchar", length), expr)
-        }
+        val trimmed = StringTrimRight(expr)
```
Comment (srowen) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: Why does the needTrim check go away here?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 3]
```diff
@@ -161,9 +162,20 @@ object CharVarcharUtils extends Logging {
 
   private def paddingWithLengthCheck(expr: Expression, dt: DataType): Expression = dt match {
     case CharType(length) =>
-      StringRPad(stringLengthCheck(expr, dt, needTrim = false), Literal(length))
-
-    case VarcharType(_) => stringLengthCheck(expr, dt, needTrim = false)
+      StaticInvoke(
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:165]: `StringRPad`'s gen-ed code is pretty short, as `s"$str.rpad($len, $pad)"`. Does `StaticInvoke` help reduce gen-ed code?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java (hunk 1 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:27]: Ok, maybe I can create a benchmark test to verify this

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 3 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:165]: It not just a single rad but,

StringRPad(If(GreaterThan(Length(StringTrimRight(expr)), Literal(length)), raiseError("char", length),StringTrimRight(expr))))

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 2 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: the `needTrim` is determined by the caller, one is read-side(!needTrim) and the other is write-side. Now the read-side logic never goes here so it is removed

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 3 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:165]: Oh I see. Then we may need a benchmark to show the difference.

**Comment by yaooqinn** [comment]
cc @cloud-fan @maropu

**Comment by yaooqinn** [comment]
# reslut from master

the performance itself does not change much, because the purpose here is to prevent queries from fallback to non-codegen
```logtalk
================================================================================================
Char Varchar Read Side Perf
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read char with length 20:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read char with length 20 wholestage off            4830           5393         796         20.7          48.3       1.0X
read char with length 20 wholestage on             1693           1755          42         59.1          16.9       2.9X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read varchar with length 20:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------
read varchar with length 20 wholestage off           4763           4884         171         21.0          47.6       1.0X
read varchar with length 20 wholestage on            1802           1888          59         55.5          18.0       2.6X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read char with length 40:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read char with length 40 wholestage off            4685           4718          46         21.3          46.9       1.0X
read char with length 40 wholestage on             1858           1904          39         53.8          18.6       2.5X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read varchar with length 40:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------
read varchar with length 40 wholestage off           4730           4749          27         21.1          47.3       1.0X
read varchar with length 40 wholestage on            1738           1831          99         57.5          17.4       2.7X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read char with length 60:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read char with length 60 wholestage off            4936           4984          68         20.3          49.4       1.0X
read char with length 60 wholestage on             1787           1835          46         55.9          17.9       2.8X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read varchar with length 60:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------
read varchar with length 60 wholestage off           4642           4740         138         21.5          46.4       1.0X
read varchar with length 60 wholestage on            1694           1755          48         59.0          16.9       2.7X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read char with length 80:                 Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
read char with length 80 wholestage off            4709           4736          38         21.2          47.1       1.0X
read char with length 80 wholestage on             1787           1869          76         56.0          17.9       2.6X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
read varchar with length 80:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
--------------------------------------------------------------------------------------------------------------------------
read varchar with length 80 wholestage off           4613           4721         153         21.7          46.1       1.0X
read varchar with length 80 wholestage on            1757           1814          59         56.9          17.6       2.6X


================================================================================================
Char Varchar Write Side Perf
================================================================================================

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write char with length 20:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write char with length 20 wholestage off           7312           7439         180          1.4         731.2       1.0X
write char with length 20 wholestage on            5957           6332         355          1.7         595.7       1.2X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write varchar with length 20:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------------------------------------
write varchar with length 20 wholestage off           9318          10265        1339          1.1         931.8       1.0X
write varchar with length 20 wholestage on            8722           9268         512          1.1         872.2       1.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write char with length 40:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write char with length 40 wholestage off          10051          10257         292          1.0        1005.1       1.0X
write char with length 40 wholestage on            8534           8917         316          1.2         853.4       1.2X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write varchar with length 40:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------------------------------------
write varchar with length 40 wholestage off          13713          14250         759          0.7        1371.3       1.0X
write varchar with length 40 wholestage on           12070          12609         516          0.8        1207.0       1.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write char with length 60:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write char with length 60 wholestage off          11728          12975        1763          0.9        1172.8       1.0X
write char with length 60 wholestage on           10211          10515         256          1.0        1021.1       1.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write varchar with length 60:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------------------------------------
write varchar with length 60 wholestage off          17845          17924         113          0.6        1784.5       1.0X
write varchar with length 60 wholestage on           16338          17353        1330          0.6        1633.8       1.1X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write char with length 80:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
write char with length 80 wholestage off          14381          14656         390          0.7        1438.1       1.0X
write char with length 80 wholestage on           13991          14556         455          0.7        1399.1       1.0X

Java HotSpot(TM) 64-Bit Server VM 1.8.0_251-b08 on Mac OS X 10.16
Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
write varchar with length 80:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
---------------------------------------------------------------------------------------------------------------------------
write varchar with length 80 wholestage off          21476          22547        1515          0.5        2147.6       1.0X
write varchar with length 80 wholestage on           21092          22393         899          0.5        2109.2       1.0X
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 2 above)
Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: I'm sure you're right, but just for more info, why does that change? this sounds like an optimization and I wasn't clear why that would mean read-side doesn't trim anymore.

Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: >  I wasn't clear why that would mean read-side doesn't trim anymore.

Looks like we have a misunderstanding here, we are not removing trim logic from the read-side in this PR, just the flag itself. The trim logic was removed in the previous PR https://github.com/apache/spark/pull/31181 as a bugfix.

Comment (srowen) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: OK, I don't question it, thanks.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+package org.apache.spark.sql.catalyst.util;
+
+import org.apache.spark.unsafe.types.UTF8String;
+
+public class CharVarcharCodegenUtils {
+  private static final UTF8String SPACE = UTF8String.fromString(" ");
+
+  public static UTF8String lengthCheck(UTF8String inputStr, int limit) {
+    if (inputStr != null && inputStr.numChars() > limit) {
+      throw new RuntimeException("Exceeds varchar type length limitation: " + limit);
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:27]: `varchar` -> `char/varchar`

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+    }
+    return inputStr;
+  }
+
+  /**
+   *  Trailing spaces do not count in the length check. We don't need to retain the trailing
+   *  spaces, as we will pad char type columns/fields at read time.
+   */
+  public static UTF8String charTypeWriteCheck(UTF8String inputStr, int limit) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/util/CharVarcharCodegenUtils.java:36]: `charTypeWriteSideCheck` to be consistent.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 6]
```diff
@@ -200,69 +212,59 @@ object CharVarcharUtils extends Logging {
...
     getRawType(targetAttr.metadata).map { rawType =>
-      stringLengthCheck(expr, rawType, needTrim = true)
+      stringLengthCheck(expr, rawType)
     }.getOrElse(expr)
   }
 
   private def raiseError(typeName: String, length: Int): Expression = {
     val errMsg = UTF8String.fromString(s"Exceeds $typeName type length limitation: $length")
-    RaiseError(Literal(errMsg, StringType), StringType)
+      RaiseError(Literal(errMsg, StringType), StringType)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:221]: nit: identation is wrong

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala [hunk 7]
```diff
@@ -200,69 +212,59 @@ object CharVarcharUtils extends Logging {
...
-              GreaterThan(Length(trimmed), Literal(length)),
-              raiseError("varchar", length),
-              StringRPad(trimmed, Literal(length))))
-        } else {
-          If(GreaterThan(Length(expr), Literal(length)), raiseError("varchar", length), expr)
-        }
+        StaticInvoke(
+          classOf[CharVarcharCodegenUtils],
+          StringType,
+          "varcharTypeWriteSidePadAndCheck",
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:238]: maybe just `varcharTypeWriteSideCheck` to show the general purpose. It also does trim but we didn't mention it in the name either.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 6 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala]: removed it

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala (hunk 7 above)
Comment (yaooqinn) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/CharVarcharUtils.scala:233]: OK

**Comment by cloud-fan** [comment]
Ideally, we shouldn't backport perf improvement, but not passing `TPCDSQuerySuite` with char/varchar in table schema concerns me (see https://github.com/apache/spark/pull/31012). Since it only touches the char/varchar code path, I feel it's safe and better to have this "fix" in branch-3.1 as well. I won't treat it as a blocker though (won't fail rc1).

@HyukjinKwon what do you think?

**Comment by yaooqinn** [comment]
+1 for @cloud-fan 's decision. As far as I'm concerned, query 41 from TPCDS is a quite normal case, 3.1 generates oversized codegen classes that let it fall back to the non-codegen mode and cause huge performance regession.  Users can hit this regularly in real-world cases if they work on their pre-defined tables with char/varchar

**Comment by HyukjinKwon** [comment]
Yes, I think it's fine to port back. It's not very common case but we can think it's like a perf regression for varchar cases compared to Spark 3.1 in a way.

**Comment by cloud-fan** [comment]
thanks, merging to master/3.1!

**Comment by yaooqinn** [comment]
thanks for merging

**Comment by maropu** [comment]
late lgtm. Nice fix and thanks, @yaooqinn

**Comment by dongjoon-hyun** [comment]
So, does it happen only with CBO and new CHAR/VARCHAR combination, @cloud-fan ?

**Comment by cloud-fan** [comment]
It's unrelated to CBO. It's about the generated code is too big and stops whole-stage-codegen.

**Pull request closed without merging**